ui/dashboards/src/components/Variables/Variable.tsx

```diff
@@ -34,33 +41,53 @@ export function TemplateVariable({ name }: TemplateVariableProps) {
   return <div>Unsupported Variable Kind: ${kind}</div>;
 }
 
+/**
+ * Returns a serialized string of the current state of variable values.
+ */
+function getVariableValuesKey(v: VariableStateMap) {
+  return Object.values(v)
+    .map((v) => JSON.stringify(v.value))
+    .join(',');
+}
+
 function ListVariable({ name }: TemplateVariableProps) {
   const ctx = useTemplateVariable(name);
   const definition = ctx.definition as ListVariableDefinition;
   const { data: variablePlugin } = usePlugin('Variable', definition.spec.plugin.kind);
-
   const { setVariableValue, setVariableLoading, setVariableOptions } = useTemplateVariableActions();
   const datasources = useLegacyDatasources();
+
+  let dependsOnVariables: string[] | undefined;
+  if (variablePlugin?.dependsOn) {
+    dependsOnVariables = variablePlugin.dependsOn(definition);
+  }
+
+  const variables = useTemplateVariableValues(dependsOnVariables);
   const allowMultiple = definition?.spec.allowMultiple === true;
   const allowAllValue = definition?.spec.allowAllValue === true;
 
-  const loadOptions = useCallback(async () => {
-    if (!variablePlugin) {
-      return;
-    }
-    setVariableLoading(name, true);
-    try {
-      const { data } = await variablePlugin.getVariableOptions(definition, { datasources });
-      setVariableOptions(name, data);
-    } catch (e) {
-      console.error('Failed to load template variable options', e);
-    }
-    setVariableLoading(name, false);
-  }, [variablePlugin, definition, name, setVariableLoading, setVariableOptions, datasources]);
+  let waitToLoad = false;
+  if (dependsOnVariables) {
+    waitToLoad = dependsOnVariables.some((v) => variables[v]?.loading);
+  }
+
+  const variablesValueKey = getVariableValuesKey(variables);
+
+  const variablesOptionsQuery = useQuery(
```

Comment:
Do you think the `usePlugin` call being a separate async call is complicating this flow/logic more than it could be? If so, you could always use the plugin registry directly to combine some of the async stuff, e.g.:

```ts
const { getPlugin } = usePluginRegistry();

const pluginStuff = useQuery([ key ], async () => {
  const plugin = await getPlugin('Variable', kind);
  const options = await plugin.getVariableOptions(...);
  return { options, plugin /* etc */ };
});
```

Not sure if that would be helpful at all, but just a thought.

Reply:
Actually decided to keep it the same so I can utilize the `enabled` property on react-query. Although, I find this API really useful.